sklearn/utils/_metadata_requests.py

```diff
@@ -763,11 +797,80 @@ def route_params(self, *, caller, params):
             res[name] = Bunch()
             for _callee, _caller in mapping:
                 if _caller == caller:
-                    res[name][_callee] = router._route_params(
-                        params=params, method=_callee
+                    res[name][_callee] = self._route_warn_or_error(
+                        child=name, router=router, params=params, method=_callee
                     )
         return res
 
+    def _route_warn_or_error(self, *, child, router, params, method):
+        """Route parameters while handling error or deprecation warning choice.
+
+        This method warns instead of raising an error if the parent object
+        has set ``warn_on`` for the child object's method and the user has not
+        set any metadata request for that child object. This is used during the
+        deprecation cycle for backward compatibility.
+
+        Parameters
+        ----------
+        child : str
+            The name of the child object.
+
+        router : MetadataRouter or MetadataRequest
+            The router for the child object.
+
+        params : dict
+            The parameters to be routed.
+
+        method : str
+            The name of the callee method.
+
+        Returns
+        -------
+        dict
+            The routed parameters.
+        """
+        try:
+            routed_params = router._route_params(params=params, method=method)
+        except UnsetMetadataPassedError as e:
+            warn_on = self._warn_on.get(child, {})
+            if method not in warn_on:
+                # there is no warn_on set for this method of this child object,
+                # we raise as usual.
+                raise
+            if not getattr(router, "_is_default_request", None):
```

Comment:
Both `MetadataRouter` and `MetadataRequest` will always have a `_is_default_request`:

```suggestion
            if not router._is_default_request:
```